sdk/core/src/CommonNetworkMember/BaseNetworkMember/RegisteringService.ts

```diff
@@ -0,0 +1,79 @@
+import { ParsedOptions } from '../../shared/getOptionsFromEnvironment'
+import { IPlatformCryptographyTools } from '../../shared/interfaces'
+import { KeyOptions } from '../../dto/shared.dto'
+import { DEFAULT_DID_METHOD, ELEM_ANCHORED_DID_METHOD, ELEM_DID_METHOD } from '../../_defaultConfig'
+import { DidDocumentService, generateFullSeed, KeysService } from '@affinidi/common'
+import { AnchoringService } from './AnchoringService'
+
+export class RegisteringService {
+  constructor(
+    private readonly options: ParsedOptions,
+    private readonly platformCryptographyTools: IPlatformCryptographyTools,
+    private readonly password: string,
+    private readonly keyOptions?: KeyOptions,
+  ) {}
+
+  private get didMethod() {
+    return this.options.otherOptions.didMethod || DEFAULT_DID_METHOD
+  }
+
+  private get didGenerationMethod() {
+    return this.didMethod !== ELEM_ANCHORED_DID_METHOD ? this.didMethod : ELEM_DID_METHOD
+  }
+
+  private get isDidElemAnchored() {
+    return this.didMethod === ELEM_ANCHORED_DID_METHOD || undefined
+  }
+
+  private get passwordBuffer() {
+    return KeysService.normalizePassword(this.password)
+  }
+
+  private async buildNewDid() {
+    const fullSeed = await generateFullSeed(this.platformCryptographyTools, this.didGenerationMethod, this.keyOptions)
+    const encryptedSeed = await KeysService.encryptSeed(fullSeed, this.passwordBuffer)
+    const keysService = new KeysService(encryptedSeed, this.password)
+
+    const didDocumentService = DidDocumentService.createDidDocumentService(keysService)
+    const didDocument = await didDocumentService.buildDidDocument()
+    const did = didDocument.id
+    return { did, encryptedSeed, didDocument, fullSeed }
+  }
+
+  private async buildElemAnchoredSeed(did: string, originalFullSeed: string) {
+    const elemAnchoredSeedDraft = await generateFullSeed(
+      this.platformCryptographyTools,
+      this.didMethod,
+      this.keyOptions,
+      {
+        anchoredDid: did,
+      },
+    )
+    const originalSeed = originalFullSeed.split('++')[0]
+    const draftSeed = elemAnchoredSeedDraft.split('++')[0]
+    const elemAnchoredSeed = elemAnchoredSeedDraft.replace(draftSeed, originalSeed)
```

Comment:
We had a method `generateFullSeed` that was useful because we never needed to generate its separate parts. This method simply generates all the parts and concatenates them together.

Now we need to create a seed that would reuse an existing first part.

The solution there would be to split `generateFullSeed` method in two (e.g. one that creates a full seed from the given first part and params; and another that generates the first part and then invokes the first method; the second method would be equivalent to the existing `generateFullSeed`), and to invoke the first method here.

Not to invoke `generateFullSeed` with unneeded data, and then do some magic with string splitting and string replacement relying on the knowledge of seed internal composition.

Additionally there is `KeysService`, there is `parseDecryptedSeed` method etc; you don't have to manually split strings in order to get the original first part here.

Reply:
let's define it as a next refactoring, anyway elem anchored future requires more work and probably we don't have time to refactor all codebase here

Reply:
I do not agree that we should write a bunch of unmaintainable, unclear and brittle code in this PR and refactor it at some time later in the future, instead of writing maintainable code (which would also be 3x smaller) from the beginning.

CC @carlos-affinidi

Reply:
Having Sidetree is indeed a priority and understandably @anton-iskryzhytskyi is pushing some refactoring to the backlog. @anton-iskryzhytskyi doesn't have the same knowledge on the SDK that you have @inga-affinidi. How can we enable developers to more easily contribute on the SDK? Do we need a more in depth documentation explaining the different parts and intricacies of the SDK? @inga-affinidi please add these guidelines to the SSDLC document, so we write maintainable code from the beginning as much as possible.

Also, if you really consider that this refactor should be done right away, let's find a way to help @anton-iskryzhytskyi with them so he can focus on Sidetree. Let's not block him at this point, even if that means a new MR.

Reply:
> @anton-iskryzhytskyi doesn't have the same knowledge on the SDK that you have @inga-affinidi.

Different people have different knowledge, but they can share this knowledge during code review?

But also Anton _already_ refactored `generateFullSeed` and `parseDecryptedSeed` in this PR, so I don't see how this is about knowledge.

> Also, if you really consider that this refactor should be done right away, let's find a way to help @anton-iskryzhytskyi with them so he can focus on Sidetree.

What kind of help do you have in mind? The first comment in this thread already describes everything that could be done regarding this point.

If a code is needed, here is what I'm talking about (just one of ways of doing this, there are others):

* In `common-libs/common/src/shared/seedTools.ts`, replace the existing code:
  ```
  export const generateFullSeed = async (
    platformCryptographyTools: IPlatformCryptographyTools,
    didMethod: string,
    keyOptions?: KeyOptions,
    metadata?: Record<string, any>,
  ): Promise<string> => {
    const seed = await randomBytes(32)
    const seedHex = seed.toString('hex')

    // ... the remaining code
  }
  ```
  with
  ```
  export const generateFullSeed = async (
    platformCryptographyTools: IPlatformCryptographyTools,
    didMethod: string,
    keyOptions?: KeyOptions,
    metadata?: Record<string, any>,
  ): Promise<string> => {
    const seed = await randomBytes(32)
    const seedHex = seed.toString('hex')

    return generateFullSeedForSeedHex(platformCryptographyTools, seedHex, didMethod, keyOptions, metadata)
  }

  export const generateFullSeedForSeedHex = async (
    platformCryptographyTools: IPlatformCryptographyTools,
    seedHex: string,
    didMethod: string,
    keyOptions?: KeyOptions,
    metadata?: Record<string, any>,
  ): Promise<string> => {
    // ... the remaining code
  }
  ```

* In `RegisteringService.ts`, replace the existing code in `buildElemAnchoredSeed`
  ```
  const elemAnchoredSeedDraft = await generateFullSeed(
    this.platformCryptographyTools,
    this.didMethod,
    this.keyOptions,
    {
      anchoredDid: did,
    },
  )
  const originalSeed = originalFullSeed.split('++')[0]
  const draftSeed = elemAnchoredSeedDraft.split('++')[0]
  const elemAnchoredSeed = elemAnchoredSeedDraft.replace(draftSeed, originalSeed)
  ```
  with
  ```
  const { seedHex } = getDecryptedSeedParts(decryptedSeeds)
  const elemAnchoredSeed = await generateFullSeedForSeedHex(
    this.platformCryptographyTools,
    seedHex,
    this.didMethod,
    this.keyOptions,
    {
      anchoredDid: did,
    },
  )
  ```

> Let's not block him at this point, even if that means a new MR.

What is the purpose of this code review, then?
If the priority is to merge this anyway ASAP, without any additional changes, then we probably should just merge this PR as is, without reviewing, as soon as it stopped being a draft?

Reply:
Although now that I look at this class (and related ones) more closely, I see that the current implementation is really hard to follow and that it can be much improved.

Current logic in this new file (it didn't exist prior to this PR) for all did methods except for `ELEM_ANCHORED_DID_METHOD`:
* Generate a new seed for this did method and `keyOptions`
  * It generates a new random `seedHex` and a bunch of random `externalKeys` according to `keyOptions`
* Encrypt it
* Create a `KeysService` instance for this encrypted seed
* Create a `DidDocumentService` instance for `KeysService` (it will decrypt the seed to extract did method and return a specific implementation for this did method)
* Create a did document by calling `buildDidDocument`
* Anchor did by calling `AnchoringService` for this encrypted seed and password and didDocument and nonce and did method (`AnchoringService` is also new and didn't exist prior to this PR)
  * nonce is always zero
  * encrypted seed and password are only used to obtain the decrypted seed
  * it only uses a first part (`seedHex`) of a decrypted seed (except that instead of just using `seedHex` it converts `seed` buffer to hex string, where `seed` buffer is obtained by converting `seedHex` from hex string to buffer in `parseDecryptedSeed`)
  * it always returns `void`
* Return did and encrypted seed

But the logic becomes even weirder when did method is `ELEM_ANCHORED_DID_METHOD`:
* Generate a new seed for `ELEM_DID_METHOD` and `keyOptions`
  * It generates a new random `seedHex` and a bunch of random `externalKeys` according to `keyOptions`
* Encrypt it
* Create a `KeysService` instance for this encrypted seed
* Create a `DidDocumentService` instance for `KeysService` (it will decrypt the seed to extract did method which is always `ELEM_DID_METHOD` and return a specific implementation for it)
  * `DidDocumentService` for `ELEM_DID_METHOD` uses `KeyVault`, and `KeyVault` decrypts seed and only uses `seedHex` and `externalKeys` (which were generated at random at the first step according to `keyOptions`)
* Create a did document by calling `buildDidDocument` (it seems that it adds public parts of `externalKeys` to the resulting document)
* Anchor did by calling `AnchoringService` for this encrypted seed and password and didDocument and nonce and `ELEM_DID_METHOD` (`AnchoringService` is also new and didn't exist prior to this PR)
  * nonce is always zero
  * for `ELEM_DID_METHOD` it does not use encrypted seed or password at all
  * it only uses a first part (`seedHex`) of a decrypted seed (except that instead of just using `seedHex` it converts `seed` buffer to hex string, where `seed` buffer is obtained by converting `seedHex` from hex string to buffer in `parseDecryptedSeed`)
  * it always returns `void`
* Take did and unencrypted seed
* Generate a _new_ seed for `ELEM_ANCHORED_DID_METHOD` and `keyOptions` and additional metadata (did from the previous step)
  * It generates a new random `seedHex` and a bunch of random `externalKeys` according to `keyOptions`
* Extract `seedHex` from the original seed and from the new seed
* Perform the string replacement (`.replace` method) on the new seed, replacing its `seedHex` with the original `seedHex`
* Discard the original seed (along with its `externalKeys`), discard the new `seedHex`
* Encrypt the new seed
* Return did and the encrypted seed.

This code is very complicated and hard to follow (it took me 40 minutes just to write a preceding part). The intent is almost impossible to deduce. And this is all in a new code.

It also looks like there is a bug: did document is created for the first seed, will contain public external keys from the first seed, but the second seed contains another set of external keys while the first set is discarded. So private keys corresponding to the public keys stored in the did document (which is a part of a returned seed, and which also was anchored) do not exist anymore; and public keys corresponding to the private external keys stored in returned seed (these private keys are used to sign data in some scenarios) are not actually made public.
Maybe this is not a bug but there are no comments to indicate that this is an intended behavior, and it does not look very intended.
And this is caused by the original code smell, when we generate one seed, do a bunch of things with it, and then generate another seed entirely (replacing its `seedHex` with the one from original seed) and discard the first one.

I can only attempt to guess what we were trying to achieve here: we needed seeds of this type to be created as the ordinary elem seeds, but also to contain a did in their metadata.
This could have been achieved in other easier to understand way:
* Treat ELEM_ANCHORED_DID_METHOD in the similar fashion as ELEM_DID_METHOD during the registration; get rid of "if it is ELEM_ANCHORED_DID_METHOD, use ELEM_DID_METHOD instead" conditions in code.
  * In `DidDocumentService`, make two separate methods for different purposes: `buildDidDocument` for registration, and `getDidDocument` for other purposes (post-registration)
* Instead of handling the entire encrypted seed during registration, only handle needed parts of unencrypted seed during registration;
* Only create one seed data with all the required data (except for metadata with did)
* After did is created, create a final seed (with did if needed)
  * These methods should be in `seedTools`. Instead of `generateSeed`, that should be e.g. `generateSeedData` (does not accept any metadata yet, returns an object) and `buildSeed` (accepts generated seed data plus metadata, returns a string)